### PR TITLE
Add contributor ladder.

### PR DESCRIPTION
**Pull request by wlynch**
See https://groups.google.com/g/tekton-governance/c/4n50jeqnDnA for motivation.

This is largely based off of
https://github.com/cncf/project-template/blob/main/CONTRIBUTOR_LADDER.md.

This helps define more lower-stakes roles for getting involved with Tekton, as
well as updates existing roles to provide expectations of the roles.
Additionally, this adds a formal process to revoke permissions in
instances of inactivity or abuse.

I tried my best to keep the requirements the same as before, so this
shouldn't be introducing any new requirements for the
collaborator/maintainer roles.

Would appreciate majority review from gov-board 🙏

### CODE REVIEW

**Review by afrittoli — COMMENTED** [review]
Thank you for this!!
I couldn't finish the review yet, sharing my feedback this far. 
I will try and review the rest today.

File: process.md
Comment (wlynch) [reply process.md:146]: Agreed that voting has its own requirements (which is actually more in line with org member)

w.r.t. events - depends on the event, but in general I agree and would expect most events we run to be community events and not contributor events.

I included this because I know that [k8s runs contributor events at kubecon](https://www.kubernetes.dev/events/kcsna2021/) that are intended to be smaller scale fun events intended for contributors, and I could see us potentially doing something similar at cdCon. :)

File: process.md
Comment (wlynch) [reply process.md]: tweaked to try and better match current requirements.

File: process.md
Comment (wlynch) [reply process.md:168]: > Being part of the org is required to be able to be assigned to an issue, which is another privilege

Is this a prow restriction? [GitHub docs](https://docs.github.com/en/issues/tracking-your-work-with-issues/assigning-issues-and-pull-requests-to-other-github-users#about-issue-and-pull-request-assignees) say:

```
You can assign up to 10 people to each issue or pull request, including yourself,
anyone who has commented on the issue or pull request, anyone with write permissions
to the repository, and organization members with read permissions to the repository.
```

---

Added bit for automatic test running

Comment (afrittoli) [reply process.md:168]: Oh, interesting:

> anyone who has commented on the issue or pull request

I didn't know about it. At least being part of the org allows to assign without a prior comment :D

**Review by dibyom — COMMENTED** [review]
Thanks @wlynch This looks great! Most of my comments are small nitpicks

File: process.md
Comment (wlynch) [reply process.md]: Done.

File: process.md
Comment (wlynch) [reply process.md]: Fixed.

**Comment by wlynch** [comment]
Ping for another round of review! 🙏

File: process.md
Comment (bobcatfish) [inline process.md:107]: nit: could you add a list here that lists the roles in the ladder below?

File: process.md
Comment (bobcatfish) [inline process.md:132]: nit: community participant <-- link here plz (i know its right above but i think it helps be totally clear)

(and the same comment for the following sections as well)

File: process.md
Comment (bobcatfish) [inline process.md:135]: since this links to the community contributing guide specifically this might be a bit confusing - for the first time maybe ever i find myself suggesting that we actually remove the link 🤣  and just explicitly say the CONTRIBUTING.md in the repo?

File: process.md
Comment (bobcatfish) [inline process.md:164]: nit are these separate requirements? e.g. do i need to author 5 PRs, and review 5 PRs? or a combo (e.g. author 2, review 3)?

File: process.md
Comment (bobcatfish) [inline process.md:170]: im not quite sure about requiring folks be an org member to vote - i dont have strong feelings but it's more restrictive than our current requirements (and if we update this here, we should update the section in governance.md that you linked to as well)

File: process.md
Comment (bobcatfish) [inline process.md:179]: nit: collaborator and maintainer seem to map almost exactly to reviewer and approver - could it be more clear to use those terms instead? (np if not)

File: process.md
Comment (bobcatfish) [inline process.md:345]: probably good to define "contribution" here - i think this means not meeting the requirements for the particular role?

**Review by bobcatfish — COMMENTED** [review]
I pretty much only have a bunch of nits but overall looks great to me!

The one non-nit thing I'm wondering is around changing the election voter requirements; I don't have super strong feelings but I want to make sure we really want to make it so that less folks would be eligible to vote (if I'm understanding correctly)

File: process.md
Comment (wlynch) [reply process.md]: Added a contributions section above based on the voter eligibility requirements.

I think we want to keep this as contributions instead of requirements though. - we define requirements pretty broadly and on a yearly basis to give people leeway for contributions over time. This is basically a check to ensure that participation is happening on a regular basis rather than a burst a short period of time.

e.g. for collaborator we define a requirement as 10 PRs a year. Ideally ~1 PR / month would be better than 10 PRs in one month, then little/no engagement for the rest of the year.

Thinking over this some more made me realize that no communication would almost certainly happen before no-contributions, so tweak the numbers a bit.

Let me know what you think!

**Review by vdemeester — APPROVED** [review]

**Review by dibyom — APPROVED** [review]

File: process.md
Comment (bobcatfish) [reply process.md:120]: not important enough to block the PR on but pinging this again @wlynch

File: process.md
Comment (bobcatfish) [reply process.md:161]: @wlynch maybe we could include something here to say that contributors maybe be eligible to vote and run in elections and link to the requirements?

File: process.md
Comment (bobcatfish) [reply process.md:195]: In that case let's at least change collaborator to reviewer - i dont feel super strongly tho so if anyone else does plz chime in ( @afrittoli @vaikas @abayer @dibyom )

I kinda see what you mean about approver v. maintainer so im okay with leaving maintainer as is

**Review by bobcatfish — COMMENTED** [review]
a few more nits from me but looking great overall!

File: process.md
Comment (dibyom) [reply process.md:195]: don't have strong opinions on this. Collaborator/Reviewer both sound good to me.

File: process.md
Comment (wlynch) [reply process.md:120]: Done.

File: process.md
Comment (wlynch) [reply process.md:171]: Done.

File: process.md
Comment (wlynch) [reply process.md]: Changed collaborator to reviewer - keeping maintainer as is.

**Comment by bobcatfish** [comment]
Thanks for the updates @wlynch , looks great to me!

/approve

File: process.md
Comment (afrittoli) [reply process.md:193]: We should probably ask people for some time and eventually enforce it

**Review by afrittoli — COMMENTED** [review]
Thank you @wlynch and sorry about the delay in reviewing this 🙏 

/approve

**Comment by bobcatfish** [comment]
woooot thanks all!!

/lgtm

**Pull request merged**
